### ISSUE TITLE
Does KerasModel is supported for mixed precision quantization?

### ISSUE THREAD

**Issue by HensonMa** [intel/neural-compressor#734]
I run the following codes, but it prompts me with the error when reaching `mix_precision.fit()` as 
```
AttributeError: 'KerasModel' object has no attribute 'graph_def'.
```
```python
from tensorflow import keras
import numpy as np

def return_Sequential():

    (train_images, train_labels), (test_images, test_labels) = keras.datasets.fashion_mnist.load_data()
    train_images = train_images.astype(np.float32) / 255.0
    test_images = test_images.astype(np.float32) / 255.0

    # Create Keras model
    model = keras.Sequential([
    keras.layers.InputLayer(input_shape=(28, 28), name="input"),
    keras.layers.Reshape(target_shape=(28, 28, 1)),
    keras.layers.Conv2D(filters=12, kernel_size=(3, 3), activation='relu'),
    keras.layers.Conv2D(filters=12, kernel_size=(3, 3), activation='relu'),
    keras.layers.Conv2D(filters=12, kernel_size=(3, 3), activation='relu'),
    keras.layers.Conv2D(filters=12, kernel_size=(3, 3), activation='relu'),
    keras.layers.Conv2D(filters=12, kernel_size=(3, 3), activation='relu'),
    keras.layers.MaxPooling2D(pool_size=(2, 2)),
    keras.layers.Flatten(),
    keras.layers.Dense(10, activation="softmax", name="output")
    ])

    # Print model architecture
    model.summary()

    # Compile model with optimizer
    opt = keras.optimizers.Adam(learning_rate=0.01)
    model.compile(optimizer=opt,
                loss="sparse_categorical_crossentropy",
                metrics=["accuracy"])

    # Train model
    model.fit(x={"input": train_images}, y={"output": train_labels}, epochs=1)
    model.save("./models/saved_model")

    return

class Dataset(object):
    def __init__(self):
        (train_images, train_labels), (test_images,
                    test_labels) = keras.datasets.fashion_mnist.load_data()
        self.test_images = test_images.astype(np.float32) / 255.0
        self.labels = test_labels
        pass

    def __getitem__(self, index):
        return self.test_images[index], self.labels[index]

    def __len__(self):
        return len(self.test_images)

# Define a customized Metric function 
from neural_compressor.metric import BaseMetric
class MyMetric(BaseMetric):
    def __init__(self, *args):
        self.pred_list = []
        self.label_list = []
        self.samples = 0
        pass

    def update(self, predict, label):
        self.pred_list.extend(np.argmax(predict, axis=1))
        self.label_list.extend(label)
        self.samples += len(label) 
        pass

    def reset(self):
        self.pred_list = []
        self.label_list = []
        self.samples = 0
        pass

    def result(self):
        correct_num = np.sum(
            np.array(self.pred_list) == np.array(self.label_list))
        return correct_num / self.samples

return_Sequential()
from neural_compressor.data import DataLoader
dataset = Dataset()
dataloader = DataLoader(framework='tensorflow', dataset=dataset)

from neural_compressor.config import MixedPrecisionConfig
from neural_compressor import mix_precision
config = MixedPrecisionConfig()
q_model = mix_precision.fit(
    model='./models/saved_model',
    config=config,
    eval_dataloader=dataloader, 
    eval_metric=MyMetric()) 

# Optional, run quantized model
import tensorflow as tf
with tf.compat.v1.Graph().as_default(), tf.compat.v1.Session() as sess:
    tf.compat.v1.import_graph_def(q_model.graph_def, name='')
    out = sess.run(['Identity:0'], feed_dict={'input:0':dataset.test_images})
    print("Inference is done.")
```

#### My environment:
- python=3.8.16
- tensorflow==2.9.3
- neural compressor: directly built from repo (2023/03/21 main branch)

**Comment by HensonMa** [comment]
#### Update
- No issue when using intel-tensorflow==2.3.0. 
- Same error when using intel-tensorflow==2.9.1. (`AttributeError: 'KerasModel' object has no attribute 'graph_def'`.
- Another error during quantization when using official tensorflow==2.3.0 (`ValueError: Please install TensorFlow within version >=1.14.0 and <=2.11.0.`)

**Comment by lvliang-intel** [comment]
Hi @HensonMa,

The official TensorFlow version prior to 2.6.0 does not support the oneDNN int8 kernel. To use it, you must install intel-tensorflow as suggested in the link below.

[TensorFlow*](https://github.com/Intel-tensorflow/tensorflow), including [1.15.0 UP3](https://github.com/Intel-tensorflow/tensorflow/tree/v1.15.0up3), [1.15.0 UP2](https://github.com/Intel-tensorflow/tensorflow/tree/v1.15.0up2), [1.15.0 UP1](https://github.com/Intel-tensorflow/tensorflow/tree/v1.15.0up1), [2.1.0](https://github.com/Intel-tensorflow/tensorflow/tree/v2.1.0), [2.2.0](https://github.com/Intel-tensorflow/tensorflow/tree/v2.2.0), [2.3.0](https://github.com/Intel-tensorflow/tensorflow/tree/v2.3.0), [2.4.0](https://github.com/Intel-tensorflow/tensorflow/tree/v2.4.0), [2.5.0](https://github.com/Intel-tensorflow/tensorflow/tree/v2.5.0)

We encountered the 'KerasModel' object has no attribute 'graph_def' error while using both official TensorFlow 2.3.0 and Intel-TensorFlow 2.3.0. This error occurred before the tensorflow version check. Have you made any local changes to the INC code? 

The 'KerasModel' object has no attribute 'graph_def' error is related with keras in kers out feature not being well adapted to mix-precision, We are currently investigating and preparing a solution to fix it. Thanks for raising this issue.

**Comment by HensonMa** [comment]
Oh, I just found a strange and interesting issue: in fact I first run `return_sequential()` using higher version of TF to train and save the model to the corresponding path, and that's how I didn't encounter `'KerasModel' object has no attribute 'graph_def'` error when using either intel or official 2.3.0 TF version.  When I keep the consistency to directly run the codes above, the error is reproducible.

**Comment by zehao-intel** [comment]
> I didn't make any local changes to the Neural Compressor code, and the codes doesn't give me `'KerasModel' object has no attribute 'graph_def'` error when using Intel-TensorFlow 2.3.0, and official Tensorflow 2.3.0.
> 
> And what if using tensorflow version higher than 2.6.0 , will it lead to issues?

If you use tensorflow, of which version is higher than 2.6.0, you are expected to pass the version check. But you will still meet the issue.

**Comment by zehao-intel** [comment]
> * tensorflow==2.9.3

I have reproduced what you described.
As we know, there are two types of saved_model on TF2.x:
    1. Regular saved_model
    2. Keras saved_model
You can use the method _tf.keras.models.load_model(model)_ to load the second one. But it may fail with the first one.
That's why we use the code below as a part of our logic to decide if a model is keras model:
```python
            try:
                model = tf.keras.models.load_model(model)
                if isinstance(model, tf.keras.Model) and hasattr(model, 'to_json'):
                    return 'keras'
                return 'saved_model'
            except:
                pass
```
When you use the intel-TF 2.3.0 to load a saved_model saved with TF 2.9.3, the _'try-except'_ logic jump to _'except'_ because an error is raised with _'tf.keras.models.load_model(model)'_.
```shell
*** json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```
Therefore, in the later logic, the model is recognized as saved_model that is not generated by keras API.
It will be wrapped by our class 'TensorflowSavedModelModel' instead of 'KerasModel'. And the former one supports 'graph_def' method. That's why it works in this case.

**Comment by HensonMa** [comment]
Thanks！！ Now we know why we encountered these issues.

**Comment by zehao-intel** [comment]
Yes. 
And we are actively reaching the owner of the feature 'keras in keras out' to see if we can give a native solution of mix-precision for Keras model.

**Comment by lvliang-intel** [comment]
Hi @HensonMa,

The issue has been resolved through the pull request #753 on https://github.com/intel/neural-compressor. You can now try again using the master branch. Please let's know if there are any further issues. Thank you very much for raising this issue.
